Add EIP-5388: Token-Gated HTTP Endpoints (#5388)
ethereum/EIPs: EIPS/eip-5388.md

---
eip: 5388

> title: Token-Gated HTTPS Endpoints

Reviewer: ```suggestion
title: Token-Gated HTTP Endpoints
```

> description: Composable RESTful and Solidity interface to implement token-gated HTTP endpoints using data tokens.

Reviewer 2: I think HTTP and RESTful can be decoupled. If this EIP mandates RESTful API, the title shall reflect it. If it only mandates HTTP, then I think we can drop the `RESTful` in the description and just keep it about HTTP endpoint?

Author: ok

Reviewer 2: Cool!

I kind of like to also have a RESTFul standard, but it can be one level above HTTP, hence a separate EIP.

> description: Composable RESTful and Solidity interface to implement token-gated HTTPS endpoints usind data tokens.

Reviewer: ```suggestion
description: Composable RESTful and Solidity interface to implement token-gated HTTP endpoints using data tokens.
```

author: Tim Daubenschütz (@TimDaub)
discussions-to: https://ethereum-magicians.org/t/token-gated-https-endpoints/10205
status: Draft
type: Standards Track

> category: ERC

Reviewer 3: I *believe* the Networking category is exclusively for Ethereum protocol P2P networking stuff.  At the least, I don't think we have ever included anything in that category other than Ethereum P2P stuff and if we want to expand its scope we should discuss that.  ERC is the catch-all for basically everything that isn't Ethereum core protocol.

> category: ERC

Reviewer: I *really* feel that this should be a networking EIP, as it kinda is on the networking layer instead of the application layer.

> category: ERC

Reviewer 2: I feel it's more of a ERC... or maybe it's opening up a new category, similar to all the URI / URL efforts.

> category: ERC

Author: I'd say that it's similar to the URL efforts in that it's not directly Ethereum client related such that e.g. core devs would have to get involved - but it's also not specifically a contract interface.

> category: ERC

Author: pls EIP editors recommend the definitive category for now, thx

> category: ERC

Reviewer 2: Hey @TimDaub are you blocked on answer this this question? We could bump it into some next Editor Office Hour and check if there is a consensu

> category: ERC

Reviewer 4: This is an ERC, not a networking EIP.

created: 2022-08-01

> requires: 712

Reviewer: ```suggestion
requires: 20, 712
```

Author: I'm against adding EIP-20 as a requirement. The first sections use EIP-20 as an example for payments. But essentially, this document isn't opinionated which smart contract method is being called or if tokens are transferred as a result of it.

---

## Abstract

> This standard introduces composable RESTful HTTP and Solidity interfaces that enable paying to access HTTP endpoints using [EIP-20](./eip-20.md) tokens.

Reviewer 5: I am not sure it is about REST?
Also, why require ERC20... I think it could be more flexible...

Reviewer 5: `This standard introduces composable HTTP and Solidity interfaces that enable restricting to access HTTP endpoints using smart contracs.`

> This standard introduces composable RESTful HTTPS and Solidity interfaces that e.g. enable paying to access HTTPS endpoints using [EIP-20](./eip-20.md) tokens.

Reviewer: ```suggestion
This standard introduces composable RESTful HTTP and Solidity interfaces that enable paying to access HTTP endpoints using [EIP-20](./eip-20.md) tokens.
```

## Motivation

> According to RFC 7231, HTTPS status `402 Payment Required` is reserved for future use. But the future is now as Ethereum enables users to verify their identity with a server using public key infrastructure.

Reviewer: ```suggestion
According to RFC 7231, HTTP status code `402 Payment Required` is reserved for future use. But the future is now as Ethereum enables users to verify their identity with a server using public key infrastructure.
```

> The Ethereum community has expressed the need for a composable method of token-gating HTTPS endpoints. In this document, we outline a protocol implementable by HTTPS-providing servers to accept [EIP-20](./eip-20.md) payments prior to allow access to an authorized endpoint.

Reviewer: ```suggestion
The Ethereum community has expressed the need for a composable method of token-gating HTTP endpoints. In this document, we outline a protocol implementable by HTTP servers to accept [EIP-20](./eip-20.md) payments prior to allowing access to an authorized endpoint.
```

## Specification

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

> Upon the user's initial request, a HTTPS server implementing token-gating via EIP-5388 responds with a `402 Payment Required` status code:

Reviewer: ```suggestion
Upon the user's initial request, an HTTP server implementing token-gating via this EIP responds with a `402 Payment Required` status code:
```

- Its body must include a valid JSON string that complies with the Solidity Contract ABI Specification for calling contracts.

> - It must include a custom header `X-EIP-5388-CONTRACT-ADDRESS` that suggests the contract's address a call must be directed to.

Reviewer 2: ```suggestion
- It **MUST** include a custom header `X-EIP-5388-CONTRACT-ADDRESS` that suggests the contract's address a call must be directed to.
```

> - It must include a custom header `X-EIP-5388-CONTRACT-ADDRESS` that suggests the contract's address a call must be directed to.

Reviewer 5: The use of the `X-` prefix, while very common is actually considered useless (if not harmful..) so I think it would be ok to skip it in favor `EIP-5388`.

Reviewer 5: I think it is important to make sure this address is compatible beyond Ethereum mainnet by adding the chain id.

Author: > The use of the `X-` prefix, while very common is actually considered useless (if not harmful..) so I think it would be ok to skip it in favor `EIP-5388`.

Can u please add a reference that explains why the `X- prefix` is harmful/useless?

Author: > I think it is important to make sure this address is compatible beyond Ethereum mainnet by adding the chain id.

Makes sense, will add it.

Reviewer: > Can u please add a reference that explains why the X- prefix is harmful/useless?

`X-` headers are for non-standardized headers. So if I have an Application A that needs to return a custom header (for some reason), I should prefix it with `X-`.

This is a standard. You can (and should) drop the `X-`.

Reviewer 5: > Can u please add a reference that explains why the X- prefix is harmful/useless?

There is a good SO thread there and my good friend Peter wrote an RFC a while back https://www.rfc-editor.org/rfc/rfc6648

https://stackoverflow.com/questions/3561381/custom-http-headers-naming-conventions

> - It must include a custom header `X-EIP-5388-ABI-ENCODED-INPUT` that suggests the call signature the user must invoke at `X-EIP-5388-CONTRACT-ADDRESS` to gain access to the endpoint.

Reviewer 2: ```suggestion
- It **MUST** include a custom header `X-EIP-5388-ABI-ENCODED-INPUT` that suggests the call signature the user must invoke at `X-EIP-5388-CONTRACT-ADDRESS` to gain access to the endpoint.
```

> - It must include a custom header `X-EIP-5388-ABI-ENCODED-INPUT` that suggests the call signature the user must invoke at `X-EIP-5388-CONTRACT-ADDRESS` to gain access to the endpoint.

Reviewer 2: I feel the spec also lacks a part in solidity interface for what smart contract it will be to respond to a end point request

Author: not sure I understand what you mean, would you mind elaborating?

Reviewer 6: I think what @xinbenlv is asking is probably which are these contracts that these endpoints might be dependent on, as suggested in his changes. Is that it?

### Example Response

> ```

Reviewer: ```suggestion
```shell
```

HTTP/2.0 402 Payment Required
Date: Mon, 01 Aug 2022 13:37:00 GMT
Content-Type: application/json
X-EIP-5388-CONTRACT-ADDRESS: 0x005241438cAF3eaCb05bB6543151f7AF894C5B58
X-EIP-5388-ABI-ENCODED-INPUT: 23b872dd00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007b

"{\"constant\":false,\"inputs\":[{\"name\":\"_from\",\"type\":\"address\"},{\"name\":\"_to\",\"type\":\"address\"},{\"name\":\"_value\",\"type\":\"uint256\"}],\"name\":\"transferFrom\",\"outputs\":[{\"name\":\"success\",\"type\":\"bool\"}],\"type\":\"function\"}"
```

> ### Authorization via EIP-712 Signature

Author: ```suggestion
### Authorization via [EIP-712](./eip-712) Signature
```

> After successfully calling the suggested Ethereum contract with the appropriate inputs, for a user request that containing a validly signed message for endpoint authorization, the server must allow access to the bespoke endpoint.

Reviewer: How about `said endpoint` instead of `the bespoke endpoint`?

The signature is complaint with [EIP-712](./eip-712.md):

```js
keccak256(abi.encodePacked(

>   hex"1901",

Reviewer: Typo?

Author: why?

Reviewer 2: @TimDaub , is `hex"1901"` a legitimate javascript code? if not, I'd suggest updating the code snippet declaration to what language it refers to, or update it to a valid js

		),
		keccak256(bytes(version)),
		chainid,
		keccak256(bytes(endpoint)),
		keccak256(bytes(method))
	)
);
```

where `endpoint` must represent the full URL, e.g. `https://ethereum.org/api/v1/cutedoge` and `method` an existing HTTP method.

```js
{
  "types": {
    "EIP712Domain": [
      {
        "name": "version",
        "type": "string"
      },
      {
        "name": "chainId",
        "type": "uint256"
      },
      {
        "name": "endpoint",
        "type": "string"
      },
      {
        "name": "method",
        "type": "string"
      }
    ],
    "Invocation": [
      {
        "name": "contract",
        "type": "address"
      },
      {
        "name": "input",
        "type": "string"
      }
    ],
    "primaryType": "Invocation",
    "domain": {
      "version": version,
      "chainId": chainid,
      "endpoint": endpoint,
      "method": method
    },
    "message": {
      "contract": contract,
      "input": input
    }
  }
}
```

The resulting hexadecimal-encoded signature must be included in a user's request to the endpoint as the `Bearer` value of the `Authorization` header. An example:

> ```

Reviewer: ```suggestion
```shell
```
There needs to be *some* syntax highlighting.

GET /index.html
Authorization: Bearer 0xabc...
```

Upon validating the signature and cross-checking it with the mandated on-chain interaction, a server must allow access to the appointed endpoint's resource.

## Rationale

> There is no rationale related this standard.

Reviewer: ```suggestion
TBD
```

## Backwards Compatibility

> There are no considerations related to the backwardd compatability of this standard.

Reviewer: ```suggestion
No backward compatibility issues were found.
```

## Security Considerations

> There are no security considerations related directly to the implementation of this standard.

Reviewer: ```suggestion
Needs discussion.
```
